Chapter4 (#10)
aicjanelia/BioImagingAI: docs/4-architectures.qmd

## Neural network architectures

In computer vision, multiple neural network architectures were designed for various visual tasks. These architectures take as input an image, which often has multiple channels - in the case of natural images there are three input channels for red/green/blue. The neural networks process the image and output various quantities depending on the task. For example, in the case of object recognition, the networks output a single probability vector to indicate the likelihood that each possible object (cat, dog etc) is in the image. 

> ![Alexnet architecture, adapted from [@krizhevsky2012imagenet]. This early neural network first demonstrated the capabilities of deep learning when trained on large datasets.](4-architectures/alexnet.PNG){#fig-alexnet width=90%}

Reviewer: Re: Figures 1 and 2. For those people who have never seen any architecture diagram before, they may not grasp the meaning of these figures. Adding a sentence or two of context for those individuals would be helpful. Alternatively, would you prefer to move these figures deeper into the chapter where the concepts are explained more in-depth?

![Convolutional operation with padding from this [github](https://github.com/vdumoulin/conv_arithmetic). This illustrates the creation of a single channel output from a single channel input. In practice, both inputs and outputs have multiple channels and all combinations of input and output need to be calculated and summed accordingly.](4-architectures/same_padding_no_strides.gif){#fig-convpad width=40%}

::: {.callout-note} 
If the kernel size *K* is odd and you set the `padding=K//2` (floor[K/2]) and `stride=1` as  shown in @fig-convpad, you get a **channel** of units that is the same size as the input.
:::

A convolutional layer operates under two main assumptions: 1) the computation only requires local features that are within the spatial extent of the filter operation; and 2) it is not necessary to perform different computations at different positions in the image, and thus the same filter operation can be convolutionally applied across all positions in the image. When these assumptions are acceptable, a convolutional layer can reduce the number of parameters substantially compared to linear layers.

> Taking our example from above, let’s estimate the number of parameters with filters/kernels of size 3 by 9 by 9 pixels, where 3 is the number of input channels and 9 is the size in pixel space, the “kernel size”. The number of input and output images are called “channels”, similar to the red/green/blue channels for RGB images. If we define the layer to have 6 output channels, this requires 6 of these kernels resulting in 1458 parameters in the kernels, along with 6 bias terms, resulting in 1464 parameters in total. As you can see, the number of parameters now is independent of the size of the input in pixels, and a dramatic reduction from the nearly 1 billion parameters for the dense linear layer example above. 

Reviewer: The explanation of "channels" would be especially helpful if it were included earlier in the text.

> Taking our example from above, let’s estimate the number of parameters with filters/kernels of size 3 by 9 by 9 pixels, where 3 is the number of input channels and 9 is the size in pixel space, the “kernel size”. The number of input and output images are called “channels”, similar to the red/green/blue channels for RGB images. If we define the layer to have 6 output channels, this requires 6 of these kernels resulting in 1458 parameters in the kernels, along with 6 bias terms, resulting in 1464 parameters in total. As you can see, the number of parameters now is independent of the size of the input in pixels, and a dramatic reduction from the nearly 1 billion parameters for the dense linear layer example above. 

Reviewer: Please break this into intermediate steps.

print('shape of filter W: ', layer.conv.weight.data.shape)
print('shape of b: ', layer.conv.bias.data.shape)

# define input images
n_images = 8
X = torch.zeros((n_images, 3, 100, 100)) # n_images x c_in x h x w

# evaluate network
layer.eval()
with torch.no_grad():
    y = layer(X)

print('input shape: ', X.shape)
print('output shape: ', y.shape)
```

The activations of a single convolutional layer, as described, have a receptive field size equivalent to its kernel size: each activation only receives information from pixels within the kernel size. However, objects in images are often larger than the kernel size. To increase the amount of spatial information used for computation, pooling layers are introduced between convolutional layers. A pooling layer consists of a sliding-window operation, just like the convolution layer, but in this case a maximum or average operation is computed within the window, independently for each input channel. To reduce the size of the image, this convolutional operation is applied with a “stride”: to downsample the image by a factor of two, we may use a pooling size of 2 with a stride also set to 2, like in the example below. 

> ![Illustration of max-pooling with a kernel size of 2 and stride of 2, from [here](https://github.com/dvgodoy/PyTorchStepByStep).](4-architectures/pooling.png){#fig-pool width=90%}

Reviewer: ```suggestion
![Illustration of max-pooling with a kernel size of 2 and stride of 2 on an image with one channel, one z-slice and 6 pixels in x and y. The output only has 3 pixels in x and y due to the pooling step. Example from from [here](https://github.com/dvgodoy/PyTorchStepByStep).](4-architectures/pooling.png){#fig-pool width=90%}
```

Code to implement max-pooling with a kernel size of 2 and a stride of 2:

```{python}
# max pooling operation
pool_layer = nn.MaxPool2d(kernel_size=2, stride=2)
y = pool_layer(X)

print('input shape: ', X.shape)
print('output shape: ', y.shape)
```

> If we apply several convolutional and pooling layers, we end up with an output which is smaller in resolution than the input. We can see example activations across convolutional, pooling and linear layers in a network trained to classify MNIST [@lecun1998mnist] digits:

Reviewer: ```suggestion
If we apply several convolutional and pooling layers, we end up with an output which is smaller in resolution than the input. We can see example activations across convolutional, pooling and linear layers in a network trained to classify MNIST [@lecun1998mnist] digits (a dataset of handwritten numbers often used to test ML ideas):
```

> ![Activations across convolutional and pooling layers from this [demo](https://adamharley.com/nn_vis/cnn/2d.html). Lighter blue indicates a higher activation.](4-architectures/mnist.PNG){#fig-mnist}

Reviewer: ```suggestion
![Each layer of a small, illustrative network where the output channels of each layer are shown as images. Lighter blue indicates higher activation at each layer. The final output layer represents the probability of the input image corresponding to a value between 0 and 9. In this example, the network correctly identifies the input as the number 3. To see this in action, checkout the interactive demonstration [here](https://adamharley.com/nn_vis/cnn/2d.html).](4-architectures/mnist.PNG){#fig-mnist}
```

 
The downsampling results in a loss of fine spatial information. To recover this information, the output of the convolutional layers in the encoder is concatenated with the activations from the decoder at each spatial scale using skip connections (“copy” operation in @fig-unet). This preserves the higher resolution details, which is important for precise segmentations and pixel-wise predictions.

Conventional u-nets [@ronneberger2015u] have two convolutional layers per block and a small kernel size of 3 in each layer. The downsampling after each block is often set to a factor of 2. Because the kernel size is small, the only way to have large receptive field sizes is through several downsampling blocks. If we want the network to learn complicated tasks across many diverse images, then we need it to have a large “capacity”. This can be achieved by adding more weights to the network, for example by increasing the number of channels in the convolutional layers and/or by adding more convolutional layers in each block and/or by increasing the number of downsampling and upsampling stages [@stringer2021].

### Vision transformers

> Vision transformers are modern architectures that are replacing convolutional networks in many applications. They are not as parameter-efficient as convolutional neural networks - for example the Cellpose segmentation u-net has 6.6 million parameters while ViT-H (“vision-transformer-huge”) has 632 million parameters [@dosovitskiy2020image]. They are still much more efficient than dense linear layers due to special architecture choices (see below), and they introduce a new type of operation called “self-attention”. Transformers avoid overfitting this large set of parameters through training on very large amounts of data. Even though they have many more parameters than standard convolutional networks, they are not too much slower because most of the operations within the transformer are matrix multiplications which are fast on newer GPUs with tensor cores (e.g. from [nvidia](https://www.nvidia.com/en-us/data-center/tensor-cores/)). With more parameters, they have a larger capacity than standard convolutional neural networks to learn from large training datasets.

Reviewer: This is potentially silly, but it is possible the Janelia lawyers will not appreciate linking to a specific product. Is it possible to refer to "tensor cores" without linking to NVIDIA directly?

![Segment Anything model [@kirillov2023segment], illustration from [@pachitariu2025cellpose]. This architecture differs from a standard transformer in the decoding module, which is necessary for outputting segmentations of images. This model requires inputs in the form of point clicks, box outlines or textual instructions.](4-architectures/sam.PNG){width=60% #fig-sam}

SAM is trained as a promptable model, meaning a user can specify an object for segmentation with a click or a bounding box and the model will produce a predicted object in that region. However, we often want automated segmentations in which we do not have to click on everything in the image. Thus, multiple groups have adapted the SAM decoder to predict auxiliary variables, which we will discuss next as we learn about loss functions. Other innovations in transformers have adapted the architecture to be more suited to the spatial layout of images, such as the Swin transformer [@liu2021swin], which has also been adapted for biological applications [@zhang2025swincell; @achard2025cellseg3d].

## Loss functions

> In a standard image classification network, the output is a vector with  length equal to the number of classes in the task. Each of the entries in this vector represents the predicted probability of the class, and the predicted label for the image is chosen as the index of the vector with the largest entry. For each training image we have a ground-truth label for the class. How close the network matches the label is the loss, which is defined as a function between the vector output of the network and the ground-truth label. A lower loss means we matched the ground-truth data better. The gradients of the network are computed automatically via back-propagation, and an optimizer is specified to modify the parameters in order to minimize the loss (described more below).

Reviewer: This is great to include!

For classification, we are evaluating predicted probabilities, and so we need to take the output from the network and convert it to a probability across classes that sums to one. For this a softmax operation is performed, defined as $p_c(x) = e^{x_{c}} / (\sum_{i=1}^C e^{x_{i}})$. The cross-entropy loss is a standard loss function for classification, and for comparing probability distributions generally. This loss maximizes the predicted probability of the true class, achieving its minimum value of 0 when the probability of the true class is predicted to be one:

$$ \ell(x, y) = - \sum_{c=1}^C \log [p_c(x)] y_c.$$

In segmentation and biological classification tasks, as mentioned before, the output is often the same size as the input in pixels and the loss is computed per-pixel. There will thus be multiple outputs of this size, each one corresponding to a class, like the entries in the vector for overall image classification. In the original u-net paper, the loss was defined using two classes, "not cell" and "cell" [@ronneberger2015u]. 

> ![Binary cross-entropy loss, computed from the output of a u-net trained to classify cell/not cell (predicted cell probability).](4-architectures/bceloss.PNG){#fig-bce width=70%}

Reviewer: A walk-through of this figure in the text would add a lot of context for it.

In pytorch, the softmax and cross-entropy loss are combined into a single function; for two-class prediction the function is [`nn.BCEWithLogitsLoss`](https://docs.pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html), and for multi-class prediction the function is [`nn.CrossEntropyLoss`](https://docs.pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html). Here is example code with cell/not-cell prediction with one convolutional layer:

```{python}
# declare layer
layer = ConvolutionalLayer(c_in=3, c_out=1, K=9)
print('shape of filter W: ', layer.conv.weight.data.shape)
print('shape of b: ', layer.conv.bias.data.shape)

# define input images
n_images = 8
X = torch.zeros((n_images, 3, 50, 50))
# define random cell probabilities
y_true = (torch.randn((n_images, 1, 50, 50)) > 0.5).float()

# define loss function
loss_fn = nn.BCEWithLogitsLoss()

# compute network prediction and loss
y = layer(X)

print('input shape: ', X.shape)
print('output shape: ', y.shape)
print('target shape: ', y_true.shape)

# evalulate loss function
loss = loss_fn(y, y_true)

print('loss = ', loss.item())
```

> To create segmentations for each cell, a threshold is defined on the cell probability and any pixels above the threshold that are connected to each other are formed into objects. This threshold is defined using a validation set - images that are not used for training or testing - to help ensure the threshold generalizes to the held-out test images. The predicted segmentations with this loss function often contain several merges, because cells can often touch each other and the connected components of the image will combine the touching cells into single components. 

Reviewer: these concepts are first explained in Section 4.3. Perhaps consider moving that explanation up or cross-reference that section here.

Author: Sorry I'm not sure if there is a repeat here, but after I push my changes, please make another suggestion

![Additional loss functions for segmentation.](4-architectures/auxloss.PNG){width=80% #fig-aux}

## Training neural networks

Now that we have defined a loss function, we want to minimize the loss $\ell$ by updating the weights in the network. For a problem like segmentation, we will need images with ground-truth segmentation labels - this labeling can be done using tools like [Ilastik](https://www.ilastik.org/), [ImageJ](https://imagej.net/software/imagej/), [Paintera](https://github.com/saalfeldlab/paintera), or [Napari](https://napari.org/stable/). Once ground-truth labeling is done on some images, training can be attempted. We will want to use most of the ground-truth labeled images for training, making up a training set, and leave a small subset (like 10%) for testing the performance of the network. For some algorithms, we may also need a validation set for setting post-processing parameters like the cell probability threshold, in which case we can reserve 10-15% of the training set images for validation.

> Minimizing the loss is done via gradient descent, in which the weights are updated by the gradient of the loss with respect to each parameter, scaled by the learning rate $\alpha$:

Reviewer: Please break this down into intermediate steps.

$$ 
\begin{align*}
\vec{v}_t &= \alpha \, d L(\vec{w}_t) / d \vec{w}_t \\
   \vec{w}_t &= \vec{w}_{t-1} - \vec{v}_t
\end{align*}
$$

![Gradient descent illustration, from [here](https://compneuro.neuromatch.io/tutorials/W1D5_DeepLearning/student/W1D5_Tutorial1.html).](4-architectures/grad_descent.gif){#fig-gd}
 
Moving the parameters in the negative direction of the gradient reduces the loss for the given images or data points over which the loss is computed. We could compute the loss and gradients over all images in the training set, but this would take too long so in practice the loss is computed in batches of a few to a few hundred images - the number of images in a batch is called the *batch size*. The optimization algorithm for updating the weights in batches is called stochastic gradient descent (SGD). This is often faster than full-dataset gradient descent because it updates the parameters many times on a single pass through the training set (called “epoch”). Also, the stochasticity induced by the random sampling step in SGD effectively adds some noise in the search for a good minimum of the loss function, which may be useful for avoiding local minima.

> It can also be beneficial to include momentum, with some value $\beta$ between zero and one, which pushes weight updates along the same direction they have been updating in the past. The updated version of $\vec{v}$ in this case is

Author: FYI this is the classical reason for momentum https://en.wikipedia.org/wiki/Gradient_descent#Momentum_or_heavy_ball_method

$$ \vec{v}_t = \beta \vec{v}_{t-1} + \alpha\, \frac{d L(\vec{w}_t)}{d \vec{w}_t} $$

Different weights in the network may have differently scaled gradients, and thus a single learning rate may not work well. The Adam optimizer uses a moving average of both the first and second moment of the gradient for rescaling the weight updates while including a momentum term [@kingma2014adam]. This optimizer works better than standard SGD in many cases, and requires less fine-tuning to find good hyperparameter values. In addition to using an optimizer like Adam, it may be helpful to use a learning rate schedule which reduces the learning rate towards the end of training to enable smaller steps for fine-tuning the final weights [@loshchilov2016sgdr]. Sometimes a validation set is used to re-instantiate the best weights, as evaluated on the validation set, before a decrease in the learning rate [@prechelt1998automatic].

> During fitting it is important to monitor the training loss and the validation loss. With an appropriate learning rate that is not too large, the training loss should always decrease. The loss on held-out validation data should also ideally decrease over training. If not, then the network is overfitting to the training set: the weights are becoming specifically tuned for the training set examples and no longer generalize to held-out data.

Reviewer: This is great to mention.

![Example training loss and validation loss across epochs.](4-architectures/trainloss.png){#fig-trainloss width=40%}

> To avoid overfitting, regularization is often used. Most commonly in computer vision problems, weight decay will be used for regularization, which is closely-related to L2 regularization. This operation reduces the weights by a small fraction $\lambda$ at each optimization step: 

Reviewer: We don't think the intended audience is likely familiar with regularization at all. We think this is a good change to generalize this explanation.
